### PR TITLE
cmd/snap: wait while inhibition file is present

### PR DESCRIPTION
**Pull request by zyga**
Inhibition files are stored in /run/snapd/inhibit and are named after
the snap name. Presence of a non-empty file stalls "snap run" from
starting additional application processes or hooks.

A non-empty file contains a hint about what is going on that doesn't
require a round trip to snapd.

Zenity and text-based UI is provided while inhibition is in place.
Currently it shows a simple progress bar and doesn't talk to snapd.
Eventually it can ask snapd for inhibition details via a REST API.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by pedronis** [comment]
Given that there can be reboots during snap refreshes is /run appropriate ?

**Comment by zyga** [comment]
This is based on https://github.com/snapcore/snapd/pull/8566

File: cmd/snap/inhibit.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:4]: I think it's unlikely this file was originally written in 2016 :smile:

File: cmd/snap/inhibit.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	switch hint {
+	case runinhibit.HintInhibitedForRefresh:
+		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
+	default:
+		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+	}
+}
+
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_SOCKET") != ""
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:42]: this should also check that the `zenity` command exists, otherwise this will confuse the user I think getting an error about zenity not being found

File: cmd/snap/inhibit.go (hunk 1 above)
Comment (zyga) [reply cmd/snap/inhibit.go:4]: Yeah, that's a copy-paste.

File: cmd/snap/inhibit.go (hunk 2 above)
Comment (zyga) [reply cmd/snap/inhibit.go:42]: Good point. I'll look at what snapd userd does, as we have some of this "GUI" logic already. It's a very rough early branch for feedback so I was not going after the full tested robust implementation yet.

File: cmd/snap/inhibit.go (hunk 1 above)
Comment (zyga) [reply cmd/snap/inhibit.go:4]: Fixed.

File: cmd/snap/inhibit.go (hunk 2 above)
Comment (zyga) [reply cmd/snap/inhibit.go:42]: I've done the minimal work required to handle missing zenity.

**Comment by zyga** [comment]
This is failing on:
```
error: open /var/lib/snapd/inhibit/test-snapd-sh.lock: permission denied
```
I need to see what's going on.

**Comment by zyga** [comment]
This is failing on:
```
error: open /var/lib/snapd/inhibit/test-snapd-sh.lock: permission denied
```
I need to see what's going on.

**Comment by zyga** [comment]
Reproduced.

I'm trying to understand what caused this to start failing.

**Comment by zyga** [comment]
This should now pass. I've explained what's going on in the last two pathces.

File: cmd/snaplock/runinhibit/inhibit.go [hunk 3]
```diff
@@ -113,7 +113,8 @@ func Unlock(snapName string) error {
 // It returns the current, non-empty hit if inhibition is in place. Otherwise
```
Comment (anonymouse64) [inline cmd/snaplock/runinhibit/inhibit.go:113]:
```suggestion
// It returns the current, non-empty hint if inhibition is in place. Otherwise
```

File: cmd/snap/inhibit.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	defer cmd.Process.Signal(os.Interrupt)
+	// Wait for zenity to terminate and store the error code.
+	// The way we invoke zenity --progress makes it wait forever.
+	// so it will typically be an external operation.
+	go func() {
+		zenityDied <- cmd.Wait()
+	}()
+
+	// Every second check if the inhibition file is still present.
+	ticker := time.NewTicker(1 * time.Second)
```
Comment (mvo5) [inline cmd/snap/inhibit.go:84]: My gut feeling is that 0.5s or even 0.2s is also fine here just to improve the "snappiness"  :) I guess we cannot select() or something on that file?

File: cmd/snap/inhibit.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case err := <-zenityDied:
```
Comment (mvo5) [inline cmd/snap/inhibit.go:89]: What will happen if the user closes (or kills) zenity? We would continue and try to run the snap? I guess that's okay, we are telling the user the refresh is in progress so I guess if things don't work the user was warned?

File: cmd/snap/inhibit.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	fmt.Fprintf(Stdout, "%s\n", i18n.G("please wait..."))
+	// TODO: display a spinner or something like that
+	// Every second check if the inhibition file is still present.
+	ticker := time.NewTicker(1 * time.Second)
```
Comment (mvo5) [inline cmd/snap/inhibit.go:114]: I would also check slightly more often here.

File: cmd/snap/inhibit.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	fmt.Fprintf(Stdout, "%s\n", i18n.G("please wait..."))
```
Comment (mvo5) [inline cmd/snap/inhibit.go:111]: You could use something like:
```
pb := progress.MakeProgressMeter()
pb.Spin(inhibitMessage(snapName, hint)) + i18n.G("please wait...")))
```
in cmd_wait.go we use 100ms as the spin interval.

Using the spinner here would also allow you to get rid of the headless flow because the spinner will DTRT when running without a terminal and just display a quiet (nop) progress bar.

File: cmd/snap/inhibit.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+			}
+			if hint == runinhibit.HintNotInhibited {
+				break loop
+			}
+		}
+	}
+	return nil
+}
+
+func headlessFlow(snapName string) error {
```
Comment (mvo5) [inline cmd/snap/inhibit.go:133]: Using a pb.ProgressMeter this could go away

File: cmd/snap/inhibit.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("snap package cannot be used")
+
+	// Run zenity with a progress bar.
+	// TODO: while we are waiting ask snapd for progress updates and send those
+	// to zenity via stdin.
+	zenityDied := make(chan error, 1)
+	cmd := exec.Command(
```
Comment (mvo5) [inline cmd/snap/inhibit.go:61]: I think we some sort of "XXX" or "TODO" here to use a dbus API in the long run to allow integration with the native desktop environment.

**Review by mvo5 — DISMISSED** [review]
LGTM, thanks for working on this. Some suggestions inline but no blockers (and everything can be fixed in followups)

File: cmd/snap/inhibit.go (hunk 5 above)
Comment (zyga) [reply cmd/snap/inhibit.go:89]: IIRC we do *not* run the app in this case.

File: cmd/snap/inhibit.go (hunk 4 above)
Comment (zyga) [reply cmd/snap/inhibit.go:84]: I'd keep it at 1 for now and follow up with inotify. Our aggressive checks often eat an entire core, and on slower systems that's a lot of waste. In practice refresh is slow enough that this doesn't matter. Users will read the message far far slower than the granularity of one second.

File: cmd/snap/inhibit.go (hunk 8 above)
Comment (zyga) [reply cmd/snap/inhibit.go:133]: This will be used by services and I'd rather explicitly do nothing, rather than rely on a third party library to do the right thing.

Comment (mvo5) [reply cmd/snap/inhibit.go:133]: This is not a third party library, it's our own code :) And it's written/documented exactly for cases like this where we need to either run on a interactive or non-interactive terminal, i.e. to avoid having to have different code-paths.

File: cmd/snap/inhibit.go (hunk 4 above)
Comment (mvo5) [reply cmd/snap/inhibit.go:84]: It's fine either way. I just not entirely agree that checking 2 times a second (stating a file) or even 5 times a second is hogging core and producing lot's of waste. It should not make any measurable difference on even very old systems. But it's something we can always tweak so it's fine.

File: cmd/snap/cmd_run.go [hunk 10]
```diff
@@ -202,6 +203,16 @@ func (x *cmdRun) Execute(args []string) error {
 	snapApp := args[0]
 	args = args[1:]
 
+	// If the snap is inhibited from being used then postpone running it until
+	// that condition passes. Inhibition UI can be dismissed by the user, in
+	// which case we don't run the application at all.
+	if features.RefreshAppAwareness.IsEnabled() {
+		snapName, _ := snap.SplitSnapApp(snapApp)
+		if err := waitWhileInhibited(snapName); err != nil {
```
Comment (pedronis) [inline cmd/snap/cmd_run.go:211]: I think we shouldn't do this if the app is a service I think, we have other mechanisms to deal with that, and I fear this code might end up into confusing situation with services

File: cmd/snap/inhibit.go (hunk 5 above)
Comment (pedronis) [inline cmd/snap/inhibit.go:89]: all the wait loops are very similar,  maybe the can be factored as some function:

waitInhbititUnlock(snapName,  errCh  <-chan error) error

**Review by pedronis — COMMENTED** [review]
did a quick pass, some comments/questions

File: cmd/snap/cmd_run.go (hunk 10 above)
Comment (zyga) [reply cmd/snap/cmd_run.go:211]: Do you mean we should just not implement inhibition at all?

Comment (pedronis) [reply cmd/snap/cmd_run.go:211]: we can discuss, the problem is that this arbitrary waiting if a service gets this far will mess up with systemd timeouts and notion of failures. We have code to disable services while we refresh already, or not disable them if the snap ask us not to

Comment (zyga) [reply cmd/snap/cmd_run.go:211]: I agree with the rationale, I'm almost done with the other, related, branch and I should be able to address all the feedback tomorrow.

**Comment by zyga** [comment]
I will work on this now. My current plan is to:

1) have services ignore run inhibition logic entirely
2) use desktop notifications, via the agent, to display a graphical notification (and cancel it when it's done)
3) print text to tty if the user session agent is gone (for use cases where people ssh into a server and maybe don't have the agent running)

**Comment by zyga** [comment]
Step 2) from the list in the preceding comment is now done. This should be worked on now, as there are no more blockers.

**Comment by jqadev** [comment]
@zyga is this issue the last blocker to enable refresh app awareness? 
Without refresh app awareness usage of apps like Chromium and IntelliJ IDEA at work is problematic so I wish every snap user to have this feature available soon and working correctly.

**Comment by zyga** [comment]
@jqadev I don't know. I'm not in sync with current state of snapd development.

**Comment by jqadev** [comment]
@mvo5 @pedronis @anonymouse64 what is the status of this blocker of the core desktop snap apps' feature i.e. refresh app awareness? 

Looking at the board https://github.com/snapcore/snapd/projects/3 this is the last issue.
The discussion here https://forum.snapcraft.io/t/wip-refresh-app-awareness/10736/60 is not very active.

It is very annoying being at work and using Chromium and Intellij Idea that are broken after every sudden snap update and then need to be restarted manually.

How many years will it pass before you turn on this feature again?

**Comment by pedronis** [comment]
@jqadev 

These issues are still active: https://forum.snapcraft.io/t/wip-refresh-app-awareness/10736/50 . (I don't have a recent update from @jamesh on them.)

We are also working on the snap side of this, to let the snap themselves know they are being held and give them some control. This relates to a larger feature where snaps can prevent, to an extent, their and other refreshes that disrupt their functionality. The feature interacts with this one in ways that require both to be stabilized before it makes sense to make them both not experimental at this point.

About timings, I hope it will be months not years.

**Comment by piyaphart90** [comment]
@zyga 

_Sent from my SM-G998B using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

File: cmd/snap/inhibit.go (hunk 7 above)
Comment (stolowski) [reply cmd/snap/inhibit.go]: Done.

File: cmd/snap/inhibit.go (hunk 9 above)
Comment (stolowski) [reply cmd/snap/inhibit.go:64]: Added.

File: cmd/snap/cmd_run.go (hunk 10 above)
Comment (stolowski) [reply cmd/snap/cmd_run.go]: I've addressed this point.

File: cmd/snap/inhibit.go (hunk 5 above)
Comment (stolowski) [reply cmd/snap/inhibit.go]: Done.

File: cmd/snap/inhibit.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	// TODO: use a dbus API to allow integration with native desktop environment.
+	cmd := exec.Command(
+		"zenity",
+		// [generic options]
+		"--title="+zenityTitle,
+		// [progress options]
+		"--progress",
+		"--text="+inhibitMessage(snapName, hint),
+		"--pulsate",
+		"--no-cancel",
```
Comment (jhenstridge) [inline cmd/snap/inhibit.go:72]: Might it be worth leaving the cancel button here? From what I can see, this should cause Zenity to exit with a non-zero exit code, and cancel the execution of the app. I worry about filling the user's screen with progress dialogs that they can't get rid of.

**Review by jhenstridge — CHANGES_REQUESTED** [review]
I've left a few change suggestions below.  It'd be good to move away from Zenity in future (maybe a desktop notification managed by session agent?), but that part seems good for now.

File: cmd/snap/inhibit.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+func isInteractiveConsole() bool {
+	return isStdoutTTY
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("snap package cannot be used")
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap/inhibit.go]: The general plan as discussed is to replace zenity with notifications in next steps, but also land this PR, and I hope this zenity code will go away soon. I'm not sure if we have any less-alarming options atm and if it's worth doing if this code is doomed anyway?

[the remark this answers is not in the document]
Comment (jhenstridge) [reply cmd/snap/inhibit.go]: Maybe something like "Snap package waiting for update" would work? I realise that this is an intermediate design: I'd just like to make sure it doesn't turn people off who happen to beta test the feature.

**Review by jhenstridge — APPROVED** [review]
Changing review to approve, since the actual bugs have been addressed. It's mostly just UI text issues now.

File: cmd/snap/export_test.go [hunk 13]
```diff
@@ -408,3 +411,19 @@ func MockOsChmod(f func(string, os.FileMode) error) (restore func()) {
...
+	}
+}
+
+func MockIsLocked(f func(snapName string) (runinhibit.Hint, error)) (restore func()) {
+	old := isLocked
+	isLocked = f
+	return func() {
+		isLocked = old
+	}
+}
```
Comment (pedronis) [inline cmd/snap/export_test.go:429]: missing end new line

File: cmd/snap/inhibit.go [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("Snap package cannot be used")
```
Comment (pedronis) [inline cmd/snap/inhibit.go:52]: I agree with @jhenstridge on this

File: cmd/snap/inhibit.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+		return zenityFlow(snapName, hint)
+	}
+	// terminal and headless
+	return textFlow(snapName, hint)
+}
+
+var isLocked = runinhibit.IsLocked
+
+var waitInhibitUnlock = func(snapName string, errCh <-chan error) error {
+	// Every second check if the inhibition file is still present.
```
Comment (pedronis) [inline cmd/snap/inhibit.go:120]: this is now every 0.5s

File: cmd/snap/inhibit.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintf(Stderr, "%s", err)
```
Comment (pedronis) [inline cmd/snap/inhibit.go:128]: do we get here if zenity is dismissed?

File: cmd/snap/cmd_run.go [hunk 17]
```diff
@@ -464,10 +475,20 @@ func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
...
+		if err := maybeWaitWhileInhibited(snapName); err != nil {
+			return err
+		}
+	}
+
 	return x.runSnapConfine(info, app.SecurityTag(), snapApp, "", args)
 }
 
 func (x *cmdRun) snapRunHook(snapName string) error {
+	if err := maybeWaitWhileInhibited(snapName); err != nil {
```
Comment (pedronis) [inline cmd/snap/cmd_run.go:488]: as is this would stop us from running the gate-auto-refresh hook,  we need to think a bit, also we probably don't want the   the graphical UI for hooks??

File: cmd/snap/inhibit.go [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+			}
+			break loop
+		case <-ticker.C:
+			// A second has elapsed, let's check again.
+			hint, err := isLocked(snapName)
+			if err != nil {
+				return err
+			}
+			if hint == runinhibit.HintNotInhibited {
```
Comment (pedronis) [inline cmd/snap/inhibit.go:137]: now that we have  HintInhibitedGateRefresh  we probably need more complex behaviors, we probably want to wait without displaying anything while the lock is HintInhibitedGateRefresh and display things only once it's really HintInhibitedForRefresh

**Review by pedronis — CHANGES_REQUESTED** [review]
there are open questions and things to adjust now that we landed some of the refresh control apps code with its changes to the run inhibit lock usage

File: cmd/snap/inhibit.go (hunk 16 above)
Comment (stolowski) [reply cmd/snap/inhibit.go:128]: Yes, closing zenity window results in exit 1, so we wouldn't wait anymore and run the app. Related to comment https://github.com/snapcore/snapd/pull/7700#discussion_r494061909

File: cmd/snap/cmd_run.go (hunk 17 above)
Comment (stolowski) [reply cmd/snap/cmd_run.go]: You're right, removed (no waiting for hooks).

File: cmd/snap/inhibit.go (hunk 18 above)
Comment (stolowski) [reply cmd/snap/inhibit.go]: Done by waiting for InhibitedForRefresh first in case it is InhibitedGateRefresh (after checking it's not HintNotInhibited first), then waiting for it to become NotInhibited.
But having second thoughts, I wonder if we need to explicitly wait for HintInhibitedForRefresh.

Comment (pedronis) [reply cmd/snap/inhibit.go]: are we avoiding displaying anything until  InhibitedForRefresh ? that was the point of asking to split the wait, otherwise indeed I suppose is not needed

Comment (stolowski) [reply cmd/snap/inhibit.go]: Yes, I made changes to waitWhileInhibited not to display antything until InhibitedForRefresh.

File: cmd/snap/inhibit.go [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, nil)
+	pb.Finished()
+	return err
+}
+
+func waitWhileInhibited(snapName string) error {
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:101]: it would be slightly easier to follow the flow from snap run if this function definition came first in the file

File: cmd/snap/inhibit.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+	if hint == runinhibit.HintNotInhibited {
+		return nil
+	}
+
+	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+	// and then either unlocks it or changes to HintInhibitedForRefresh (see
+	// gateAutoRefreshHookHandler in hooks.go).
+	// waitInhibitUnlock will return also on HintNotInhibited.
+	if err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh, nil); err != nil {
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:115]: so the flow here if the lock is held by snapd to run the gate-auto-refresh hook is that we wait for that hook to finish and either set the inhibit lock to unlocked, or to set it to HintInhibitedForRefresh. Is the expectation that the gate-auto-refresh hook will run in a "short" amount of time? because otherwise if that hook takes a while to process, `snap run` will be stuck here displaying nothing to the user, and only then after the hook is done (at which point maybe the lock is even unlocked if the refresh isn't happening) do we progress and display any indication that the process is blocked waiting for something to happen either graphically with zenity or textually. 

Basically, my concern is about the ordering if we shouldn't also display zenity while waiting for this condition as well as the condition for the refresh hook no longer being held

File: cmd/snap/cmd_run_test.go [hunk 21]
```diff
@@ -31,8 +32,11 @@ import (
 
 	"gopkg.in/check.v1"
 
+	main "github.com/snapcore/snapd/cmd/snap"
 	snaprun "github.com/snapcore/snapd/cmd/snap"
```
Comment (anonymouse64) [inline cmd/snap/cmd_run_test.go:36]: why is this imported twice ?

File: cmd/snap/cmd_run_test.go [hunk 22]
```diff
@@ -205,6 +209,130 @@ func (s *RunSuite) TestSnapRunAppIntegration(c *check.C) {
...
+	})
+	defer restorer()
+
+	c.Assert(runinhibit.LockWithHint("snapname", runinhibit.HintInhibitedForRefresh), check.IsNil)
+	c.Assert(os.MkdirAll(dirs.FeaturesDir, 0755), check.IsNil)
+	c.Assert(ioutil.WriteFile(features.RefreshAppAwareness.ControlFile(), []byte(nil), 0644), check.IsNil)
+
+	var called bool
+	restore := main.MockWaitInhibitUnlock(func(snapName string, waitFor runinhibit.Hint, errCh <-chan error) error {
+		called = true
```
Comment (anonymouse64) [inline cmd/snap/cmd_run_test.go:233]: should this function check the value of `waitFor` and the number of times it is called? it seems like this function will be called twice, first for the HintInhibitedForRefresh, and then for HintNotInhibited in the textual / zenity flows

File: cmd/snap/cmd_run_test.go [hunk 23]
```diff
@@ -205,6 +209,130 @@ func (s *RunSuite) TestSnapRunAppIntegration(c *check.C) {
...
+		execArgs = args
+		execEnv = envv
+		return nil
+	})
+	defer restorer()
+
+	var called bool
+	restore := main.MockWaitInhibitUnlock(func(snapName string, waitFor runinhibit.Hint, errCh <-chan error) error {
+		called = true
+		return nil
```
Comment (anonymouse64) [inline cmd/snap/cmd_run_test.go:273]: this should return an error and also do `c.Errorf("WaitInhibitUnlock should not have been called")`

File: cmd/snap/cmd_run_test.go [hunk 24]
```diff
@@ -205,6 +209,130 @@ func (s *RunSuite) TestSnapRunAppIntegration(c *check.C) {
...
+	})
+	defer restorer()
+
+	c.Assert(runinhibit.LockWithHint("snapname", runinhibit.HintInhibitedForRefresh), check.IsNil)
+	c.Assert(os.MkdirAll(dirs.FeaturesDir, 0755), check.IsNil)
+	c.Assert(ioutil.WriteFile(features.RefreshAppAwareness.ControlFile(), []byte(nil), 0644), check.IsNil)
+
+	var called bool
+	restore := main.MockWaitInhibitUnlock(func(snapName string, waitFor runinhibit.Hint, errCh <-chan error) error {
+		called = true
```
Comment (anonymouse64) [inline cmd/snap/cmd_run_test.go:315]: same as for the hook test case

File: cmd/snap/inhibit.go [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:51]: is the intention that we do not want to test the zenity flow because it will very shortly be replaced with a D-Bus implementation? as it is I don't see any tests exercising the zenity flow logic ?

**Review by anonymouse64 — COMMENTED** [review]
some questions, also I think we are still missing some unit tests for the zenity flow?

File: cmd/snap/inhibit.go (hunk 25 above)
Comment (stolowski) [reply cmd/snap/inhibit.go:51]: Yes, zenity will go away soon.

File: cmd/snap/inhibit.go (hunk 20 above)
Comment (stolowski) [reply cmd/snap/inhibit.go:115]: This is addressing Samuele's comment https://github.com/snapcore/snapd/pull/7700#discussion_r659741283 about not displaying anything while waiting for the hook. In theory the hook is expected to call 'snapctl refresh --hold|--proceed' (which itself is quick), but of course it may have arbitrary logic and falls into general hook timeout limit.

File: cmd/snap/inhibit.go (hunk 18 above)
Comment (anonymouse64) [reply cmd/snap/inhibit.go]: > we probably want to wait without displaying anything while the lock is HintInhibitedGateRefresh 

@pedronis but why though? as Pawel pointed out, it could take an arbitrary amount of time for that hook to finish and proceed to HintInhibitedForRefresh in which case the user sees nothing, if we refactored the flow (I guess perhaps how the flow was originally ?) then we could display something that blocks waiting for both cases to change and thus have less cases of user confusion wondering why nothing is happening when they run their snap

Comment (pedronis) [reply cmd/snap/inhibit.go]: I'll comment on this. I understand the concerns. I still think this is the baseline behavior we want. On which can improve later as needed.

File: cmd/snap/inhibit.go [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+loop:
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintf(Stderr, "%s", err)
+			}
+			break loop
+		case <-ticker.C:
+			// A second has elapsed, let's check again.
```
Comment (pedronis) [inline cmd/snap/inhibit.go:141]: half a second

File: cmd/snap/inhibit.go [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+
+	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+	// and then either unlocks it or changes to HintInhibitedForRefresh (see
+	// gateAutoRefreshHookHandler in hooks.go).
+	// waitInhibitUnlock will return also on HintNotInhibited.
+	if err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh, nil); err != nil {
+		return err
+	}
+
```
Comment (pedronis) [inline cmd/snap/inhibit.go:118]: at this point the lock could be already again runinhibit.HintNotInhibited  if the hook didn't block things, so we shouldn't display the zenity or text UI in that case but return like we do in 106, so it seems waitInhibitUnlock should return whether it hit no inhibit or a different waitFor condition

File: cmd/snap/inhibit.go (hunk 20 above)
Comment (pedronis) [reply cmd/snap/inhibit.go:115]: the issue is that the hook can either say don't proceed or proceed, it's a confusing experience from my POV to show a notification just about the fact that we are running the hook to check whether to proceed and then a different one if we are really inhibited, and we can't really say that we are waiting for the snap to be refreshed if it won't be refreshed either. So under the assumption that the hook is quick (that's under the control of the people creating snaps) this experience seems reasonable. A more refined experience would be to put a notification up if the hook is taking too long, but wait a small reasonable amount of time before doing that. This can be done as an enhancement later on top of the baseline here.

**Review by pedronis — COMMENTED** [review]
did another pass, some considerations

File: cmd/snap/inhibit.go (hunk 20 above)
Comment (anonymouse64) [reply cmd/snap/inhibit.go:115]: @pedronis that explanation makes sense, I agree the best case scenario in this case is that we only show a popup notification if the gate hook takes "too long"

File: cmd/snap/inhibit.go (hunk 27 above)
Comment (stolowski) [reply cmd/snap/inhibit.go:118]: Done.

File: cmd/snap/inhibit.go (hunk 25 above)
Comment (anonymouse64) [reply cmd/snap/inhibit.go:51]: not sure if I mentioned my opinion earlier, but at this stage I would much rather just merge this PR with zenity so that it can be merged and then open smaller followups which remove zenity and replace it with D-Bus etc so we can make more progress without having to re-review this same PR again

**Comment by anonymouse64** [comment]
@stolowski I don't see any changes here since I last reviewed?

**Comment by stolowski** [comment]
> @stolowski I don't see any changes here since I last reviewed?

Uh, sorry, I pushed to wrong (my) remote instead of zyga's 🤦 . Changes should be visible now.

**Comment by pedronis** [comment]
@stolowski if I merge master into this I see failing tests in cmd/snap:

----------------------------------------------------------------------
FAIL: cmd_run_test.go:1796: RunSuite.TestWaitWhileInhibitedTextFlow

cmd_run_test.go:1814:
    c.Check(s.Stdout(), check.Equals, "snap package cannot be used now: gate-refresh\n")
... obtained string = ""
... expected string = "snap package cannot be used now: gate-refresh\n"

cmd_run_test.go:1817:
    c.Check(meter.Finishes, check.Equals, 1)
... obtained int = 0
... expected int = 1

cmd_run_test.go:1818:
    c.Check(meter.Labels, check.DeepEquals, []string{"please wait..."})
... obtained []string = []string(nil)
... expected []string = []string{"please wait..."}
... Difference:
...     []string[0] != []string[1]


OOPS: 573 passed, 1 FAILED
--- FAIL: Test (13.93s)
FAIL
exit status 1


Also I don't know if it's expected that a small window appears quickly during them?

**Comment by stolowski** [comment]
> @stolowski if I merge master into this I see failing tests in cmd/snap:
> 
> FAIL: cmd_run_test.go:1796: RunSuite.TestWaitWhileInhibitedTextFlow
> 
> cmd_run_test.go:1814:
> c.Check(s.Stdout(), check.Equals, "snap package cannot be used now: gate-refresh\n")
> ... obtained string = ""
> ... expected string = "snap package cannot be used now: gate-refresh\n"
> 
> cmd_run_test.go:1817:
> c.Check(meter.Finishes, check.Equals, 1)
> ... obtained int = 0
> ... expected int = 1
> 
> cmd_run_test.go:1818:
> c.Check(meter.Labels, check.DeepEquals, []string{"please wait..."})
> ... obtained []string = []string(nil)
> ... expected []string = []string{"please wait..."}
> ... Difference:
> ... []string[0] != []string[1]
> 
> OOPS: 573 passed, 1 FAILED
> --- FAIL: Test (13.93s)
> FAIL
> exit status 1
> 
> Also I don't know if it's expected that a small window appears quickly during them?

It seems mocking is missing somewhere as I can't reproduce it in headless environment. Will fix, thanks.

**Comment by stolowski** [comment]
@pedronis the problem of missing mocking affecting graphical session should be fixed now.

**Review by pedronis — APPROVED** [review]
lgtm as an intermediate step before we switch away from zenity

File: cmd/snap/inhibit.go [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintf(Stderr, "%s", err)
+			}
+			break loop
```
Comment (anonymouse64) [inline cmd/snap/inhibit.go:143]: basically from my other comment I think this should read:
```suggestion
			if err != nil {
				return false, err
			}
			break loop
```

File: cmd/snap/cmd_run_test.go [hunk 29]
```diff
@@ -1583,3 +1716,106 @@ func (s *RunSuite) TestRunGdbserverNoGdbserver(c *check.C) {
...
+		if called == 1 {
+			errCh <- fmt.Errorf("boom")
+		}
+		return runinhibit.HintInhibitedForRefresh, nil
+	})
+	defer restore()
+
+	notInhibited, err := snaprun.WaitInhibitUnlock("some-snap", runinhibit.HintNotInhibited, errCh)
+	c.Assert(err, check.IsNil)
+	c.Check(notInhibited, check.Equals, false)
```
Comment (anonymouse64) [inline cmd/snap/cmd_run_test.go:1768]: this bit of the unit test confuses me, shouldn't if we have an error returned on the errCh also then have an error returned from WaitInhibitUnlock too? Otherwise snap run will just output `boom` on stderr and then try to run the snap anyways no?

**Review by anonymouse64 — COMMENTED** [review]
sorry, I have one more question about how the errCh works with waitInhibitUnlock

File: cmd/snap/cmd_run_test.go (hunk 29 above)
Comment (stolowski) [reply cmd/snap/cmd_run_test.go:1768]: Yes, but that was the idea , this relates to e https://github.com/snapcore/snapd/pull/7700/files#r659738457

File: cmd/snap/inhibit.go [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	_, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, nil)
+	pb.Finished()
+	return err
+}
+
+var isLocked = runinhibit.IsLocked
+
+var waitInhibitUnlock = func(snapName string, waitFor runinhibit.Hint, errCh <-chan error) (notInhibited bool, err error) {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap/inhibit.go:136]: Added.

File: cmd/snap/inhibit.go (hunk 28 above)
Comment (stolowski) [reply cmd/snap/inhibit.go]: Agreed, it's more readable (though it should be false, nil). Applied, thanks.

File: cmd/snap/cmd_run_test.go (hunk 29 above)
Comment (anonymouse64) [reply cmd/snap/cmd_run_test.go:1768]: ah okay sorry I missed that, this makes sense in that case. This is one of the pitfalls of having PR's open for so long haha

File: cmd/snap/inhibit.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/snap/inhibit.go:136]: thanks for that

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for the last changes
